File: packages/modules/ImsMedia/tests/native/service/src/com/android/telephony/imsmedia/lib/libimsmedia/config/TextConfigTest.cpp
/**
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <TextConfig.h>
#include <gtest/gtest.h>

using namespace android::telephony::imsmedia;
// RtpConfig
const int32_t kMediaDirection = RtpConfig::MEDIA_DIRECTION_NO_FLOW;
const android::String8 kRemoteAddress("0.0.0.0");
const int32_t kRemotePort = 10000;
const int8_t kDscp = 0;
const int8_t kRxPayload = 100;
const int8_t kTxPayload = 100;
const int8_t kSamplingRate = 8;

// RtcpConfig
const android::String8 kCanonicalName("name");
const int32_t kTransmitPort = 10001;
const int32_t kIntervalSec = 3;
const int32_t kRtcpXrBlockTypes = 0;

// TextConfig
const int32_t kCodecType = TextConfig::TEXT_T140_RED;
const int32_t kBitrate = 100;
const int8_t kRedundantPayload = 102;
const int8_t kRedundantLevel = 3;
const bool kKeepRedundantLevel = true;

class TextConfigTest : public ::testing::Test
{
public:
    RtcpConfig rtcp;
    TextConfig config1;
    TextConfig config2;
    TextConfig config3;

protected:
    virtual void SetUp() override
    {
        rtcp.setCanonicalName(kCanonicalName);
        rtcp.setTransmitPort(kTransmitPort);
        rtcp.setIntervalSec(kIntervalSec);
        rtcp.setRtcpXrBlockTypes(kRtcpXrBlockTypes);

        config1.setMediaDirection(kMediaDirection);
        config1.setRemoteAddress(kRemoteAddress);
        config1.setRemotePort(kRemotePort);
        config1.setRtcpConfig(rtcp);
        config1.setDscp(kDscp);
        config1.setRxPayloadTypeNumber(kRxPayload);
        config1.setTxPayloadTypeNumber(kTxPayload);
        config1.setSamplingRateKHz(kSamplingRate);
        config1.setCodecType(kCodecType);
        config1.setBitrate(kBitrate);
        config1.setRedundantPayload(kRedundantPayload);
        config1.setRedundantLevel(kRedundantLevel);
        config1.setKeepRedundantLevel(kKeepRedundantLevel);
    }

    virtual void TearDown() override {}
};

TEST_F(TextConfigTest, TestGetterSetter)
{
    EXPECT_EQ(config1.getCodecType(), kCodecType);
    EXPECT_EQ(config1.getBitrate(), kBitrate);
    EXPECT_EQ(config1.getRedundantPayload(), kRedundantPayload);
    EXPECT_EQ(config1.getRedundantLevel(), kRedundantLevel);
    EXPECT_EQ(config1.getKeepRedundantLevel(), kKeepRedundantLevel);
}

TEST_F(TextConfigTest, TestParcel)
{
    android::Parcel parcel;
    config1.writeToParcel(&parcel);
    parcel.setDataPosition(0);

    TextConfig configTest;
    configTest.readFromParcel(&parcel);
    EXPECT_EQ(configTest, config1);
}

TEST_F(TextConfigTest, TestAssign)
{
    TextConfig testConfig = config1;
    EXPECT_EQ(config1, testConfig);

    TextConfig* testConfig2 = new TextConfig(config1);
    EXPECT_EQ(config1, *testConfig2);
    delete testConfig2;
}

TEST_F(TextConfigTest, TestEqual)
{
    config2.setMediaDirection(kMediaDirection);
    config2.setRemoteAddress(kRemoteAddress);
    config2.setRemotePort(kRemotePort);
    config2.setRtcpConfig(rtcp);
    config2.setDscp(kDscp);
    config2.setRxPayloadTypeNumber(kRxPayload);
    config2.setTxPayloadTypeNumber(kTxPayload);
    config2.setSamplingRateKHz(kSamplingRate);
    config2.setCodecType(kCodecType);
    config2.setBitrate(kBitrate);
    config2.setRedundantPayload(kRedundantPayload);
    config2.setRedundantLevel(kRedundantLevel);
    config2.setKeepRedundantLevel(kKeepRedundantLevel);
    EXPECT_EQ(config2, config1);
}

TEST_F(TextConfigTest, TestNotEqual)
{
    config2.setMediaDirection(kMediaDirection);
    config2.setRemoteAddress(kRemoteAddress);
    config2.setRemotePort(kRemotePort);
    config2.setRtcpConfig(rtcp);
    config2.setDscp(kDscp);
    config2.setRxPayloadTypeNumber(kRxPayload);
    config2.setTxPayloadTypeNumber(kTxPayload);
    config2.setSamplingRateKHz(kSamplingRate);
    config2.setCodecType(kCodecType);
    config2.setBitrate(kBitrate);
    config2.setRedundantPayload(103);
    config2.setRedundantLevel(kRedundantLevel);
    config2.setKeepRedundantLevel(kKeepRedundantLevel);

    EXPECT_NE(config2, config1);
}

File: packages/modules/ImsMedia/tests/unit/src/com/android/telephony/imsmedia/TextConfigTest.java
/**
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.telephony.imsmedia;

import static com.google.common.truth.Truth.assertThat;

import android.net.InetAddresses;
import android.os.Parcel;
import android.telephony.AccessNetworkConstants.AccessNetworkType;
import android.telephony.imsmedia.RtcpConfig;
import android.telephony.imsmedia.RtpConfig;
import android.telephony.imsmedia.TextConfig;

import androidx.test.runner.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.net.InetSocketAddress;

@RunWith(AndroidJUnit4.class)
public class TextConfigTest {
    // RtcpConfig
    private static final String CANONICAL_NAME = "name";
    private static final int RTCP_PORT = 30001;
    private static final int RTCP_INTERVAL = 3;
    // TextConfig
    private static final String REMOTE_RTP_ADDRESS = "122.22.22.22";
    private static final int REMOTE_RTP_PORT = 30000;
    private static final byte DSCP = 10;
    private static final byte RX_PAYLOAD = 100;
    private static final byte TX_PAYLOAD = 100;
    private static final byte SAMPLING_RATE = 10;
    private static final int CODEC_TYPE = TextConfig.TEXT_T140_RED;
    private static final int BITRATE = 300;
    private static final byte REDUNDANT_PAYLOAD = 101;
    private static final byte REDUNDANT_LEVEL = 3;
    private static final boolean KEEP_REDUNDANT_LEVEL = true;

    private static final RtcpConfig sRtcp = new RtcpConfig.Builder()
            .setCanonicalName(CANONICAL_NAME)
            .setTransmitPort(RTCP_PORT)
            .setIntervalSec(RTCP_INTERVAL)
            .setRtcpXrBlockTypes(RtcpConfig.FLAG_RTCPXR_DLRR_REPORT_BLOCK)
            .build();

    @Test
    public void testConstructorAndGetters() {
        TextConfig config = createTextConfig();
        assertThat(config.getMediaDirection()).isEqualTo(
                RtpConfig.MEDIA_DIRECTION_SEND_RECEIVE);
        assertThat(config.getAccessNetwork()).isEqualTo(AccessNetworkType.EUTRAN);
        assertThat(config.getRemoteRtpAddress()).isEqualTo(new InetSocketAddress(
                InetAddresses.parseNumericAddress(REMOTE_RTP_ADDRESS), REMOTE_RTP_PORT));
        assertThat(config.getRtcpConfig()).isEqualTo(sRtcp);
        assertThat(config.getDscp()).isEqualTo(DSCP);
        assertThat(config.getRxPayloadTypeNumber()).isEqualTo(RX_PAYLOAD);
        assertThat(config.getTxPayloadTypeNumber()).isEqualTo(TX_PAYLOAD);
        assertThat(config.getSamplingRateKHz()).isEqualTo(SAMPLING_RATE);
        assertThat(config.getCodecType()).isEqualTo(CODEC_TYPE);
        assertThat(config.getBitrate()).isEqualTo(BITRATE);
        assertThat(config.getRedundantPayload()).isEqualTo(REDUNDANT_PAYLOAD);
        assertThat(config.getRedundantLevel()).isEqualTo(REDUNDANT_LEVEL);
        assertThat(config.getKeepRedundantLevel()).isEqualTo(KEEP_REDUNDANT_LEVEL);
    }

    @Test
    public void testParcel() {
        TextConfig config = createTextConfig();

        Parcel parcel = Parcel.obtain();
        config.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);

        TextConfig parcelConfig = TextConfig.CREATOR.createFromParcel(parcel);
        assertThat(config).isEqualTo(parcelConfig);
    }

    @Test
    public void testEqual() {
        TextConfig config1 = createTextConfig();
        TextConfig config2 = createTextConfig();
        assertThat(config1).isEqualTo(config2);
    }

    @Test
    public void testNotEqual() {
        TextConfig config1 = createTextConfig();

        TextConfig config2 = new TextConfig.Builder()
                .setMediaDirection(RtpConfig.MEDIA_DIRECTION_SEND_RECEIVE)
                .setAccessNetwork(AccessNetworkType.EUTRAN)
                .setRemoteRtpAddress(new InetSocketAddress(
                        InetAddresses.parseNumericAddress(REMOTE_RTP_ADDRESS), REMOTE_RTP_PORT))
                .setRtcpConfig(sRtcp)
                .setDscp(DSCP)
                .setRxPayloadTypeNumber(RX_PAYLOAD)
                .setTxPayloadTypeNumber(TX_PAYLOAD)
                .setSamplingRateKHz(SAMPLING_RATE)
                .setCodecType(TextConfig.TEXT_T140)
                .setBitrate(BITRATE)
                .setRedundantPayload(REDUNDANT_PAYLOAD)
                .setRedundantLevel(REDUNDANT_LEVEL)
                .setKeepRedundantLevel(KEEP_REDUNDANT_LEVEL)
                .build();

        assertThat(config1).isNotEqualTo(config2);
    }

    static TextConfig createTextConfig() {
        return new TextConfig.Builder()
                .setMediaDirection(RtpConfig.MEDIA_DIRECTION_SEND_RECEIVE)
                .setAccessNetwork(AccessNetworkType.EUTRAN)
                .setRemoteRtpAddress(new InetSocketAddress(
                        InetAddresses.parseNumericAddress(REMOTE_RTP_ADDRESS), REMOTE_RTP_PORT))
                .setRtcpConfig(sRtcp)
                .setDscp(DSCP)
                .setRxPayloadTypeNumber(RX_PAYLOAD)
                .setTxPayloadTypeNumber(TX_PAYLOAD)
                .setSamplingRateKHz(SAMPLING_RATE)
                .setCodecType(CODEC_TYPE)
                .setBitrate(BITRATE)
                .setRedundantPayload(REDUNDANT_PAYLOAD)
                .setRedundantLevel(REDUNDANT_LEVEL)
                .setKeepRedundantLevel(KEEP_REDUNDANT_LEVEL)
                .build();
    }
}
